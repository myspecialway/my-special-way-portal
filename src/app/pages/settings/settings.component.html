<form #formControl="ngForm" [formGroup]="form" novalidate>
    <div data-test-id="settings-page" class="main-content">
        <div class="container-fluid">
            <div class="header-menu-container msw-table-tools-header">
                <div class="header-table-count">
                    <div>
                        <span class="header-table-count-label">הגדרות</span>
                    </div>
                </div>
            </div>
            <div style="width: 360px; margin: 10px auto;">
                <mat-card class="selected-radius" style="height:75px">
                    <div class="table">
                        <div class="row">
                            <div class="cell">
                                <p>קוד מורה</p>
                            </div>
                            <div class="cell" style="width: 100px; padding-left: 20px"> 
                                <mat-form-field style="width: 100px; text-align: center" floatLabel="never">
                                    <input matInput #input data-test-id="teacher-code-input"
                                        formControlName="teachercodecontrol" name="teachercodecontrol"
                                        type="text" 
                                        [(ngModel)]="teachercode"
                                        placeholder="קוד"
                                    >
                                    <mat-error data-test-id="teachercode-err1" *ngIf="teacherCodeFormControl.hasError('required')">קוד מורה הינו שדה חובה</mat-error>
                                    <mat-error data-test-id="teachercode-err2" *ngIf="teacherCodeFormControl.hasError('minlength') || teacherCodeFormControl.hasError('maxlength') || teacherCodeFormControl.hasError('pattern')">נא הקש 4 ספרות</mat-error>
                                </mat-form-field> 
                            </div>
                        </div>
                    </div>
                </mat-card>
            </div>
        </div>
    </div>
</form>