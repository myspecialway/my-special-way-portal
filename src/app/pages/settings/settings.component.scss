@import '../../../assets/scss/msw-table';

.mat-card {
  height: 60px;
  padding: 0;
  margin-left: auto;
  margin-right: auto;
  vertical-align: middle;
}
div.row:first-child p {
  font-size: 18px;
}
div.row:nth-child(2) .cell p,
div.row:nth-child(3) .cell p {
  font-size: 16px;
  font-weight: 500;
}
.row {
  line-height: 0.7;
  display: table-row;
  .value {
    text-align: left;
    font-size: 14px !important;
  }
}
.table {
  width: 100%;
  display: table;
  text-align: right;
  @media screen and (max-width: 580px) {
    display: block;
  }
}
.cell {
  padding: 0 16px;
  display: table-cell;
  @media screen and (max-width: 580px) {
    padding: 2px 16px;
    display: block;
  }
}

.cell.input-container {
  width: 100px;
  padding-left: 20px;
  mat-form-field {
    width: 100px;
    text-align: center;
  }
}
.selected-radius:hover {
  border: 1px solid #8b66a7;
  border-radius: 5px;
  opacity: 1;
  height: 75px;
}
.content-container {
  width: 360px;
  margin: 10px auto;
}
