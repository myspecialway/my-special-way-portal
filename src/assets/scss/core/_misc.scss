body {
  background-color: #fafafa;
  color: $black-color;
  font-weight: 300;
}

legend {
  border-bottom: 0;
}

.serif-font {
  font-family: $font-family-serif;
}

// Prevent highlight on mobile
* {
  -webkit-tap-highlight-color: rgba(255, 255, 255, 0);
  -webkit-tap-highlight-color: transparent;
  &:focus {
    outline: 0;
  }
}

a {
  color: $link-color;
  &:hover,
  &:focus {
    color: darken($link-color, 5%);
    text-decoration: none;
  }

  &.text-info {
    &:hover,
    &:focus {
      color: darken($brand-info, 5%);
    }
  }

  & .material-icons {
    vertical-align: middle;
  }
}

.form-check,
label {
  font-size: 14px;
  line-height: 1.42857;
  color: $checkboxes-text-color;
  font-weight: 400;
}

/*           Animations              */
.animation-transition-general {
  @include transition-all($general-transition-time, $transition-linear);
}

.animation-transition-slow {
  @include transition-all($slow-transition-time, $transition-linear);
}

.animation-transition-fast {
  @include transition-all($fast-transition-time, $transition-ease);
}

.caret,
.sidebar a {
  @include transition-all($fast-transition-time, $transition-ease-in);
}
#map {
  position: relative;
  z-index: 2;
  width: 100%;
  height: calc(100vh - 70px);
  margin-top: 70px;
}

@font-face {
  font-family: 'pass';
  font-style: normal;
  font-weight: 400;
  src: url(data:application/font-woff;charset=utf-8;base64,d09GRgABAAAAAATsAA8AAAAAB2QAAQAAAAAAAAAAAAAAAAAAAAAAAAAAAABGRlRNAAABWAAAABwAAAAcg9+z70dERUYAAAF0AAAAHAAAAB4AJwANT1MvMgAAAZAAAAA/AAAAYH7AkBhjbWFwAAAB0AAAAFkAAAFqZowMx2N2dCAAAAIsAAAABAAAAAQAIgKIZ2FzcAAAAjAAAAAIAAAACAAAABBnbHlmAAACOAAAALkAAAE0MwNYJ2hlYWQAAAL0AAAAMAAAADYPA2KgaGhlYQAAAyQAAAAeAAAAJAU+ATJobXR4AAADRAAAABwAAAAcCPoA6mxvY2EAAANgAAAAEAAAABAA5gFMbWF4cAAAA3AAAAAaAAAAIAAKAE9uYW1lAAADjAAAARYAAAIgB4hZ03Bvc3QAAASkAAAAPgAAAE5Ojr8ld2ViZgAABOQAAAAGAAAABuK7WtIAAAABAAAAANXulPUAAAAA1viLwQAAAADW+JM4eNpjYGRgYOABYjEgZmJgBEI2IGYB8xgAA+AANXjaY2BifMg4gYGVgYVBAwOeYEAFjMgcp8yiFAYHBl7VP8wx/94wpDDHMIoo2DP8B8kx2TLHACkFBkYA8/IL3QB42mNgYGBmgGAZBkYGEEgB8hjBfBYGDyDNx8DBwMTABmTxMigoKKmeV/3z/z9YJTKf8f/X/4/vP7pldosLag4SYATqhgkyMgEJJnQFECcMOGChndEAfOwRuAAAAAAiAogAAQAB//8AD3jaY2BiUGJgYDRiWsXAzMDOoLeRkUHfZhM7C8Nbo41srHdsNjEzAZkMG5lBwqwg4U3sbIx/bDYxgsSNBRUF1Y0FlZUYBd6dOcO06m+YElMa0DiGJIZUxjuM9xjkGRhU2djZlJXU1UDQ1MTcDASNjcTFQFBUBGjYEkkVMJCU4gcCKRTeHCk+fn4+KSllsJiUJEhMUgrMUQbZk8bgz/iA8SRR9qzAY087FjEYD2QPDDAzMFgyAwC39TCRAAAAeNpjYGRgYADid/fqneL5bb4yyLMwgMC1H90HIfRkCxDN+IBpFZDiYGAC8QBbSwuceNpjYGRgYI7594aBgcmOAQgYHzAwMqACdgBbWQN0AAABdgAiAAAAAAAAAAABFAAAAj4AYgI+AGYB9AAAAAAAKgAqACoAKgBeAJIAmnjaY2BkYGBgZ1BgYGIAAUYGBNADEQAFQQBaAAB42o2PwUrDQBCGvzVV9GAQDx485exBY1CU3PQgVgIFI9prlVqDwcZNC/oSPoKP4HNUfQLfxYN/NytCe5GwO9/88+/MBAh5I8C0VoAtnYYNa8oaXpAn9RxIP/XcIqLreZENnjwvyfPieVVdXj2H7DHxPJH/2/M7sVn3/MGyOfb8SWjOGv4K2DRdctpkmtqhos+D6ISh4kiUUXDj1Fr3Bc/Oc0vPqec6A8aUyu1cdTaPZvyXyqz6Fm5axC7bxHOv/r/dnbSRXCk7+mpVrOqVtFqdp3NKxaHUgeod9cm40rtrzfrt2OyQa8fppCO9tk7d1x0rpiQcuDuRkjjtkHt16ctbuf/radZY52/PnEcphXpZOcofiEZNcQAAeNpjYGIAg///GBgZsAF2BgZGJkZmBmaGdkYWRla29JzKggxD9tK8TAMDAxc2D0MLU2NjENfI1M0ZACUXCrsAAAABWtLiugAA)
    format('woff');
}

input[type='password'] {
  font-family: 'pass', 'Roboto', Helvetica, Arial, sans-serif;
  font-size: 18px;
  &::-webkit-input-placeholder {
    transform: scale(0.77);
    transform-origin: 0 50%;
  }
  &::-moz-placeholder {
    font-size: 14px;
    opacity: 1;
  }
  &:-ms-input-placeholder {
    font-size: 14px;
    font-family: 'Roboto', Helvetica, Arial, sans-serif;
  }
}
