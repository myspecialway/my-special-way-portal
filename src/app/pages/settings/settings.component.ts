import { Component, OnInit } from '@angular/core';
import { SubscriptionCleaner } from '../../decorators/SubscriptionCleaner.decorator';
import { SettingService } from './services/settings.service';
import { Settings } from '../../models/settings.mode';
import { FormControl, Validators, FormGroup, FormBuilder } from '@angular/forms';

@Component({
  selector: 'app-settings',
  templateUrl: './settings.component.html',
  styleUrls: ['./settings.component.scss'],
})
export class SettingsComponent implements OnInit {
  teachercode: number;
  _settings: Settings | undefined;
  formControl: FormControl;
  form: FormGroup;
  teacherCodeFormControl: FormControl;

  @SubscriptionCleaner()
  subCollector;

  constructor(private formBuilder: FormBuilder, private settingsService: SettingService) {
    this.teachercode = 0;
  }

  ngOnInit(): void {
    this.subCollector.add(
      this.settingsService.getAll().subscribe((data) => {
        this._settings = data[0];
        if (this._settings !== undefined) {
          this.teachercode = this._settings.teachercode;
        }
      }),
    );
    this.createFormControls();
    this.createForm();
  }

  private createForm() {
    this.form = this.formBuilder.group({
      teachercodecontrol: this.teacherCodeFormControl,
    });
  }

  private createFormControls() {
    this.teacherCodeFormControl = new FormControl('', {
      validators: [
        Validators.required,
        Validators.minLength(4),
        Validators.maxLength(4),
        Validators.pattern('^[1-9]+$'),
      ],
      updateOn: 'blur',
    });
  }

  codeChanged(code: number): void {
    let updatedSettings: Settings = new Settings();
    if (this._settings !== undefined) {
      updatedSettings = { _id: this._settings._id, teachercode: code };
    }
    this.subCollector.add(
      this.settingsService.update(updatedSettings).then((data) => {
        this.teachercode = data[0].teachercode;
      }),
    );
  }
}
