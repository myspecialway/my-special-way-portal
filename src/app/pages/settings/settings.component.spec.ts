import { TestBed } from '@angular/core/testing';
import { NO_ERRORS_SCHEMA } from '@angular/core';

import { SettingsComponent } from './settings.component';
import { SettingService } from './services/settings.service';
import { Platform } from '@angular/cdk/platform';
import { Observable } from 'rxjs-compat';
import { settingsTestData } from '../../../mocks/assets/settings.mock';
import { FormsModule, ReactiveFormsModule, FormBuilder } from '@angular/forms';

describe('setting component', () => {
  let settingsServiceMock: Partial<SettingService>;

  beforeEach(async () => {
    settingsServiceMock = {
      getAll: jest.fn().mockReturnValueOnce(Observable.of(settingsTestData)),
      update: jest.fn(),
    };

    TestBed.configureTestingModule({
      imports: [ReactiveFormsModule, FormsModule],
      declarations: [SettingsComponent],
      providers: [FormBuilder, SettingService, { provide: SettingService, useValue: settingsServiceMock }, Platform],
      schemas: [NO_ERRORS_SCHEMA],
    });
  });

  it('should render component as described in snapshot', () => {
    const fixture = TestBed.createComponent(SettingsComponent);
    expect(fixture).toMatchSnapshot();
  });

  it('should update teachercode after get data from server', () => {
    const fixture = TestBed.createComponent(SettingsComponent);
    fixture.componentInstance.ngOnInit();
    expect(fixture.componentInstance.teachercode).toBe(1234);
  });

  it('should load settings from service on page load ', () => {
    (settingsServiceMock.getAll as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve();
    });
    const fixture = TestBed.createComponent(SettingsComponent);
    fixture.detectChanges();
    const serviceMock = TestBed.get(SettingService);
    expect(serviceMock.getAll).toHaveBeenCalled();
  });

  it('should update settings from service on code changed ', () => {
    (settingsServiceMock.update as jest.Mock).mockImplementationOnce(() => {
      return Promise.resolve({ data: { data: 1 } });
    });

    const fixture = TestBed.createComponent(SettingsComponent);
    fixture.detectChanges();
    fixture.componentInstance._settings = {
      _id: '123',
      teachercode: 1234,
    };
    const teachercode = fixture.componentInstance.form.controls['teachercodecontrol'];
    teachercode.setValue(2222);
    fixture.componentInstance.codeChanged(2222);

    fixture.detectChanges();
    const serviceMock = TestBed.get(SettingService);
    expect(teachercode.valid).toBeTruthy();
    expect(serviceMock.update).toHaveBeenCalled();
    expect(teachercode.value).toBe(2222);
  });
});
