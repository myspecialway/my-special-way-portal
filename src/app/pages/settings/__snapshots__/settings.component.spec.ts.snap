// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setting component should render component as described in snapshot 1`] = `
<app-settings
  formBuilder={[Function FormBuilder]}
  settingsService={[Function Object]}
  teachercode="0"
>
  <form
    novalidate=""
  >
    <div
      class="main-content"
      data-test-id="settings-page"
    >
      <div
        class="container-fluid"
      >
        <div
          class="header-menu-container msw-table-tools-header"
        >
          <div
            class="header-table-count"
          >
            <div>
              <span
                class="header-table-count-label"
              >
                הגדרות
              </span>
            </div>
          </div>
        </div>
        <div
          style="width: 360px; margin: 10px auto;"
        >
          <mat-card
            class="selected-radius"
            style="height:75px"
          >
            <div
              class="table"
            >
              <div
                class="row"
              >
                <div
                  class="cell"
                >
                  <p>
                    קוד מורה
                  </p>
                </div>
                <div
                  class="cell"
                  style="width: 100px; padding-left: 20px"
                >
                  <mat-form-field
                    floatlabel="never"
                    style="width: 100px; text-align: center"
                  >
                    <input
                      data-test-id="teacher-code-input"
                      formcontrolname="teachercodecontrol"
                      matinput=""
                      name="teachercodecontrol"
                      placeholder="קוד"
                      type="text"
                    />
                    
                    
                  </mat-form-field>
                </div>
              </div>
            </div>
          </mat-card>
        </div>
      </div>
    </div>
  </form>
</app-settings>
`;
